eksperyment_mikroserwisy.rst

```diff
@@ -0,0 +1,78 @@
+
+Docker Compose
+~~~~~~~~~~~~~~
+
+Uruchamianie wielu usług (nawet trzech - jak w przeprowadzonym teście) jest uciążliwe. Aby ułatwić proces zastosowano narzędzie Docker Compose. Skonfigurowano je tak, aby jedną komendą można było uruchomić trzy serwisy podłączone do odpowiednich sieci, z odpowiednimi adresami zależności.
+
+Wnioski
+-------
+
+* Prosta funkcjonalność testowego systemu okazała się łatwa do zrealizowania w architekturze mikrousług.
+
+* Frontend w modelu mikrousług można zrealizować następująco: każda usługa, która chce mieć interfejs graficzny WWW ma go niezależnie od innych usług (tzn ma wbudowany serwer HTTP, który serwuje strony). Jeśli jakaś usługa wewnątrz swojego frontendu chce załączyć frontend innej mikrousługi może zrobić to za pomocą iframe.
+
+* Gotowe mikrousługi można uruchamiać bez narzędzi wirtualizacji/konteneryzacji. Można też uruchamiać je z użyciem takich narzędzi. Użycie takich narzędzi jest dla usług przezroczyste. To dobra cecha, ponieważ pozwala na zmianę środowiska bez zmiany kodu usług.
+
+* Występują problemy z inicjalizacją trwałego stanu, co oznacza, że inicjalizacja stanu jest szczególnym przypadkiem działania usługi, który trzeba przewidzieć i obsłużyć (ten aspekt stosuje się do wszelkich programów, które mają trwały stan). Są dwie główne możliwości:
+
+  * Inicjalizacja jest przeprowadzana automatycznie, gdy aplikacja wykryje konieczność inicjalizacji.
+  * Inicjalizacja jest przeprowadzana przez inny kod niż zasadnicza aplikacja. Może to być inny program lub część aplikacji, która uruchamiana jest przez przekazanie do aplikacji odpowiedniego argumentu.
+
+* Korzystanie z zarządcy zależności nie wygląda na dobry pomysł. Każda usługa powinna przyjmować w swoich argumentach lokalizację innych aplikacji, od których zależy i nie korzystać z zarządcy zależności.
+
+* Konteneryzacja ułatwia uruchamianie usług. Np dzięki konteneryzacji nie ma problemu z kolidującymi wersjami oprogramowania (np A wymaga foobar-3.3.6, B wymaga foobar-3.4.1), nie ma problemu z kolizjami portów (wszystkie serwery HTTP, których jest n można uruchomić na porcie 80).
+
+* Do zarządzania zależnościami pomiędzy usługami dobrze sprawdza się DNS serwowany przez Dockera wewnątrz sieci wirtualnych. Np. uruchamiamy mikrousługę A i nazywamy ją "ms_a". Następnie, uruchamiając mikrousługę B, która korzysta z A podajemy jej adres A: http://ms_a/api/. Nie musimy wiedzieć jaki adres ma usługa A. Dzięki temu adres może być przydzielany dynamicznie.
+
+* Docker-compose znacznie ułatwia uruchamianie projektu poprzez automatyzację uruchamiania wielu powiązanych usług.
+
+* Być może Docker nie powinien obsługiwać stanowych usług. Stwierdzenie nie jest wynikiem eksperymentu, lecz artykułów wskazujących zagrożenia związane z uruchamianiem stanowych usług w kontenerach (https://thehftguy.com/2016/11/01/docker-in-production-an-history-of-failure/, https://thehftguy.com/2017/02/23/docker-in-production-an-update/). Cytaty:
```

Comment:
do tych artykułów dałabym `por.` albo `np.`

jak to się ma do używania baz danych poza swarmem?